src/boot/efi/xbootldr.c

```diff
@@ -190,22 +259,17 @@ VOID xbootldr_open(EFI_HANDLE *device, EFI_HANDLE *ret_device, EFI_FILE **ret_ro
                         continue;
 
                 hd = (HARDDRIVE_DEVICE_PATH*) node;
-                hd->PartitionNumber = found_partition_number;
-                hd->PartitionStart = found_partition_start;
-                hd->PartitionSize = found_partition_size;
-                CopyMem(hd->Signature, found_partition_signature, sizeof(hd->Signature));
+                hd->PartitionNumber = part_number;
+                hd->PartitionStart = part_start;
+                hd->PartitionSize = part_size;
+                CopyMem(hd->Signature, &part_uuid, sizeof(hd->Signature));
                 hd->MBRType = MBR_TYPE_EFI_PARTITION_TABLE_HEADER;
                 hd->SignatureType = SIGNATURE_TYPE_GUID;
         }
 
-        r = uefi_call_wrapper(BS->LocateDevicePath, 3, &BlockIoProtocol, &copy, &new_device);
-        if (EFI_ERROR(r))
-                return;
-
-        root_dir = LibOpenRoot(new_device);
-        if (!root_dir)
+        err = uefi_call_wrapper(BS->LocateDevicePath, 3, &BlockIoProtocol, &partition_path, ret_device);
+        if (EFI_ERROR(err))
                 return;
 
-        *ret_device = new_device;
-        *ret_root_dir = root_dir;
+        *ret_root_dir = LibOpenRoot(*ret_device);
```

Comment:
hmm, this means we'll return "half initialized" fields in some rror paths here, no?

usually our coding style suggests that we should either touch all output params (on success) or none (on failure).